<a href="https://colab.research.google.com/github/bs3537/DS-Unit-1-Sprint-2-Data-Wrangling-and-Storytelling/blob/master/Bhav_Copy_of_LS_DS_121_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Lambda School Data Science_

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-09-19 03:44:21--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.98.237
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.98.237|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.1’

instacart_online_gr 100%[===================>] 196.03M  16.6MB/s    in 13s     

2019-09-19 03:44:35 (14.8 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.1’ saved [205548478/205548478]



In [0]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [0]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [0]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

**First, I will open all CSV files and inspect the columns**

In [0]:
import pandas as pd
import numpy as np
aisles = pd.read_csv("aisles.csv")
aisles.head()

aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

In [0]:

departments = pd.read_csv("departments.csv")
departments.head()

department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [0]:
order_products_prior = pd.read_csv("order_products__prior.csv")

order_products_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [0]:
order_products_prior.shape

(32434489, 4)

In [0]:
order_products_train = pd.read_csv("order_products__train.csv")
order_products_train.head()

order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [0]:
order_products_train.shape

(1384617, 4)

In [0]:
orders = pd.read_csv("orders.csv")
orders.head()

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [0]:
products = pd.read_csv("products.csv")
products.head()

product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

**Now, I have opened all six csv files and will inspect their columns to see which ones have product names, product IDs, aisle Ids and number of orders**

**Order_product_prior dataframe has the target number of orders value that we want by product ID. If we can find product ID for these grocery items, we can find the number of count (add_to_cart_order value).**

In [0]:
products[['product_id', 'product_name']]

product_id                                       product_name
0               1                         Chocolate Sandwich Cookies
1               2                                   All-Seasons Salt
2               3               Robust Golden Unsweetened Oolong Tea
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...
4               5                          Green Chile Anytime Sauce
5               6                                       Dry Nose Oil
6               7                     Pure Coconut Water With Orange
7               8                  Cut Russet Potatoes Steam N' Mash
8               9                  Light Strawberry Blueberry Yogurt
9              10     Sparkling Orange Juice & Prickly Pear Beverage
10             11                                  Peach Mango Juice
11             12                         Chocolate Fudge Layer Cake
12             13                                  Saline Nasal Mist
13             14                     Fresh Scent Dishwasher Cleaner
14             15                           Overnight Diapers Size 6
15             16                      Mint Chocolate Flavored Syrup
16             17                                  Rendered Duck Fat
17             18                Pizza for One Suprema  Frozen Pizza
18             19   Gluten Free Quinoa Three Cheese & Mushroom Blend
19             20     Pomegranate Cranberry & Aloe Vera Enrich Drink
20             21                   Small & Medium Dental Dog Treats
21             22                  Fresh Breath Oral Rinse Mild Mint
22             23                             Organic Turkey Burgers
23             24  Tri-Vi-Sol® Vitamins A-C-and D Supplement Drop...
24             25            Salted Caramel Lean Protein & Fiber Bar
25             26        Fancy Feast Trout Feast Flaked Wet Cat Food
26             27  Complete Spring Water Foaming Antibacterial Ha...
27             28                                  Wheat Chex Cereal
28             29  Fresh Cut Golden Sweet No Salt Added Whole Ker...
29             30         Three Cheese Ziti, Marinara with Meatballs
...           ...                                                ...
49658       49659                            Organic Creamed Coconut
49659       49660                        Professionals Sleek Shampoo
49660       49661                                              Porto
49661       49662                       Bacon Cheddar Pretzel Pieces
49662       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...
49663       49664                     Lemon Cayenne Drinking Vinegar
49664       49665      Super Dark Coconut Ash & Banana Chocolate Bar
49665       49666                      Ginger Snaps Snacking Cookies
49666       49667           Enchilada with Spanish Rice & Beans Meal
49667       49668                     Apple Cinnamon Scented Candles
49668       49669                                   K Cup Dark Blend
49669       49670                                Beef Summer Sausage
49670       49671                               Milk Chocolate Drops
49671       49672                             Cafe Mocha K-Cup Packs
49672       49673              Stone Baked Multi Grain Artisan Rolls
49673       49674            Frozen Greek Yogurt Bars Chocolate Chip
49674       49675                Cinnamon Dolce Keurig Brewed K Cups
49675       49676                             Ultra Red Energy Drink
49676       49677                    Thick & Chunky Sloppy Joe Sauce
49677       49678                    Large Chicken & Cheese Taquitos
49678       49679                            Famous Chocolate Wafers
49679       49680                 All Natural Creamy Caesar Dressing
49680       49681            Spaghetti with Meatballs and Sauce Meal
49681       49682                                 California Limeade
49682       49683                                     Cucumber Kirby
49683       49684          Vodka, Triple Distilled, Twist of Vanilla
49684

**FINDING THE PRODUCT IDs of ABOVE GROCERY ITEMS FROM PRODUCTS DATAFRAME USING .LOC FUNCTION**

Next Step is finding the product IDs for above grocery intems using .loc method.

In [0]:
# finding the information for Banana

products.loc[products['product_name'] == 'Banana']

product_id product_name  aisle_id  department_id
24851       24852       Banana        24              4

In [0]:
#finding just product ID for banana

products.loc[products['product_name'] == 'Banana', ['product_id']]

product_id
24851       24852

In [0]:
products.loc[products['product_name'] == 'Bag of Organic Bananas']

product_id            product_name  aisle_id  department_id
13175       13176  Bag of Organic Bananas        24              4

In [0]:
products.loc[products['product_name'] == 'Organic Strawberries']

product_id          product_name  aisle_id  department_id
21136       21137  Organic Strawberries        24              4

In [0]:
products.loc[products['product_name'] == 'Organic Baby Spinach']

product_id          product_name  aisle_id  department_id
21902       21903  Organic Baby Spinach       123              4

In [0]:
products.loc[products['product_name'] == 'Organic Hass Avocado']

product_id          product_name  aisle_id  department_id
47208       47209  Organic Hass Avocado        24              4

In [0]:
products.loc[products['product_name'] == 'Organic Avocado']

product_id     product_name  aisle_id  department_id
47765       47766  Organic Avocado        24              4

In [0]:
products.loc[products['product_name'] == 'Large Lemon']

product_id product_name  aisle_id  department_id
47625       47626  Large Lemon        24              4

In [0]:
products.loc[products['product_name'] == 'Strawberries']

product_id  product_name  aisle_id  department_id
16796       16797  Strawberries        24              4

In [0]:
products.loc[products['product_name'] == 'Limes']

product_id product_name  aisle_id  department_id
26208       26209        Limes        24              4

In [0]:
products.loc[products['product_name'] == 'Organic Whole Milk']

product_id        product_name  aisle_id  department_id
27844       27845  Organic Whole Milk        84             16

**Concatenate order_products__prior and order_products__train**

In [0]:
order_products = pd.concat([order_products_prior, order_products_train])
order_products.shape

(33819106, 4)

**The concatenate function was successful and number of rows in the above merged output dataframe is a sum of the number of rows in individual dataframes. **

In [0]:
order_products.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

**MERGING THE DESIRED COLUMNS FROM DIFFERENT DATAFRAMES IN A SINGLE DATAFRAME CONTAINING COLUMNS OF INTEREST**

In [0]:

merged = pd.merge(products, order_products)
merged.head()

product_id                product_name  ...  add_to_cart_order  reordered
0           1  Chocolate Sandwich Cookies  ...                  7          0
1           1  Chocolate Sandwich Cookies  ...                  3          1
2           1  Chocolate Sandwich Cookies  ...                  4          1
3           1  Chocolate Sandwich Cookies  ...                  2          0
4           1  Chocolate Sandwich Cookies  ...                 30          0

[5 rows x 7 columns]

In [0]:
merged.describe()

product_id      aisle_id  ...  add_to_cart_order     reordered
count  3.381911e+07  3.381911e+07  ...       3.381911e+07  3.381911e+07
mean   2.557551e+04  7.121799e+01  ...       8.367738e+00  5.900617e-01
std    1.409770e+04  3.819898e+01  ...       7.139540e+00  4.918220e-01
min    1.000000e+00  1.000000e+00  ...       1.000000e+00  0.000000e+00
25%    1.351900e+04  3.100000e+01  ...       3.000000e+00  0.000000e+00
50%    2.525600e+04  8.300000e+01  ...       6.000000e+00  1.000000e+00
75%    3.793500e+04  1.070000e+02  ...       1.100000e+01  1.000000e+00
max    4.968800e+04  1.340000e+02  ...       1.450000e+02  1.000000e+00

[8 rows x 6 columns]

In [0]:
merged.count()

# Thus, there were 33819106 orders for all items

product_id           33819106
product_name         33819106
aisle_id             33819106
department_id        33819106
order_id             33819106
add_to_cart_order    33819106
reordered            33819106
dtype: int64

In [0]:
# Number of orders for Banana

merged.loc[merged['product_id'] == 24852].sum()

# I was able to slice the rows containing the attributes for item Banana using this function

product_id                                                 12209563932
product_name         BananaBananaBananaBananaBananaBananaBananaBana...
aisle_id                                                      11790984
department_id                                                  1965164
order_id                                                  838947943426
add_to_cart_order                                              2405664
reordered                                                       415166
dtype: object

In [0]:
merged.loc[merged['product_id'] == 24852].count()

product_id           491291
product_name         491291
aisle_id             491291
department_id        491291
order_id             491291
add_to_cart_order    491291
reordered            491291
dtype: int64

#### Thus, there were 491291 orders for Banana out of total 33819106

**We can find the number of orders for other items like strawberry etc. using the same function as above and entering their product id from above exercise**

## Reshape Data Section

- Replicate the lesson code
- Complete the code cells we skipped near the beginning of the notebook
- Table 2 --> Tidy
- Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [0]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

In [0]:
table1

treatmenta  treatmentb
John Smith           NaN           2
Jane Doe            16.0          11
Mary Johnson         3.0           1

In [0]:
table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [0]:
table2.index

Index(['treatmenta', 'treatmentb'], dtype='object')

In [0]:
tidy = table2.melt(id_vars='index')
tidy

# I am getting the error below despite repeated attempts

KeyError: ignored

In [0]:
flights = sns.load_dataset('flights')
flights.head()

year     month  passengers
0  1949   January         112
1  1949  February         118
2  1949     March         132
3  1949     April         129
4  1949       May         121

In [0]:
flights_pivot = pd.pivot_table(flights,index=["year"], columns=['month'])
flights_pivot.head()

passengers                       ...                                    
month    January February March April  ... September October November December
year                                   ...                                    
1949         112      118   132   129  ...       136     119      104      118
1950         115      126   141   135  ...       158     133      114      140
1951         145      150   178   163  ...       184     162      146      166
1952         171      180   193   181  ...       209     191      172      194
1953         196      196   236   235  ...       237     211      180      201

[5 rows x 12 columns]

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####